### DISCUSSION TITLE
Afinal, qual stack vamos utilizar no Hackernews clone?

### DISCUSSION THREAD

**Discussion by huogerac** [devpro-br/hackernews-clone#3] [General]
## De forma bem resumida, nossa stack será:

**Frontend:** VueJs com um framework CSS
**Backend:** ~FastAPI~ Flask, Openapi, SQLAlchemy, Alembic, Postgres
**Devops:** Docker, NGINX, Ansible, GH Actions, GH Container Registry
**Cloud:** Digital Ocean para ter que instalar tudo na mão (um pouco de Vercel/Heroku)

## Gostaram? 
No próximo encontro vamos entender mais o motivo dessas escolhas e por que não vamos usar Django (apenas desta vez)

Espero que gostem, acredito que vai ser bem na linha das respostas de voces no questionário #2

**Comment by serlus** [comment]
DUCARILHO!!! VueJs era o FW q estava curioso (provalmente por influência da Buser), então fico super animado. A escolha de Flask e OpenApi me deixa curioso, pois vi q o FastApi usa OpenApi no projeto e o Flask é um desafio grande no meu caso pois ele parece uma tela em branco q não sei por onde começar. 

Agora subir maquina na unha na DO, me deixou muito feliz, pq faz a parada na unha me faz valorizar ( e entender) as soluções prontas de mercado. Além de me ajudar pensar o q poderia ter feito em projetos q fiz em empresas passadas :see_no_evil:

**Comment by yzakius** [comment]
Gostei bastante e estou ansioso!

**Comment by diegomaranhao** [comment]
Bem legal, eu só gostaria do fastAPI no lugar do flask por "sentir" que está mais em alta, mas assim está ótimo também.

**Comment by tiagotardelli** [comment]
Acho que CSS podemos utilizar SASS, o que acham?

**Comment by huogerac** [comment]
O **Luciano Martins** iniciou uma pergunta bem interessante no group Telegram da Confraria DevPro:

> mas vai começar com flask? pq? se td mundo aki está aprendendo nas aulas do Renzo o Django?

Bom Luciano, este é um curso para todos, mas principalmente para quem já está no grupo e querendo ver algo um pouco mais além do Django e principalmente um frontend fora do django. Entende?

Flask? Bom o foco não é no micro-framework em si, o foco será fazer as APIs, como ter um frontend separado acessando esta API etc...

```

Frontend --> API Client --> API --> Services --> DB (migrations, ORM)

```

Flask para resolver a API é apenas um pedaço! A ideia principal é a gente entender mais um pouco de cada coisa neste fluxo...
Eu te entendo perfeitamente, eu mesmo era muito focado em Java, tudo que era fora disto eu ignorava completamente!

Se você acha que tem muita coisa para estudar no Django ainda, você vai poder voltar mais tarde neste conteúdo! Não se preocupe
Se não, você é mais que bem vindo! não existe pré requisitos para o que vamos falar e fazer nestes encontros!

Valeu

**Reply by Riverfount** [reply]
Olá @huogerac eu ainda sugiro, se não for muito complicado, usarmos [fastAPI](https://fastapi.tiangolo.com/), pois ela já nos fornece, de quebra, a documentação em OpenAPI sem termos que instalar qualquer extensão extra como temos que fazer tanto em Flask como em Django, o que me diz?

**Reply by huogerac** [reply]
Hey @Riverfount , legal! Eu gosto bastante do FastAPI também! E acho interessante o recurso de "ganhar" a documentação da API.

Porém, no plano de estudo vamos focar na parte "Fullstack", e sabendo mais de toda arquitetura, nos leva a pensar em formas de melhorar o desenvolvimento e autonomia dos times.

Um problema comum que observo é a equipe do frontend ficar dependente do backend, ou seja, muitas vezes o pessoal do frontend tem que esperar aquela mudança da API para poder finalizar um determinada funcionalidade/página.

Ai entra o "Design first API", uma abordagem um pouco diferente de fazer APIs, a documentação guia a implementação, logo "ganhar" a documentação como o FastAPI faz, não ajuda!
Pode até parecer estranho em um primeiro momento, mas tenho certeza que você vai gostar! Acredito que vamos ver isto na Aula 3.

Desta forma, Flask e sua caracteristica de ser micro, pelado! podemos plugar apenas o que precisamos! Assim conseguimos configurar o OpenAPI de uma forma que permite gerar rotas vindo da documentação (e não o contrário)

**Reply by huogerac** [reply]
@Riverfount @diegomaranhao 

Com o bate papo do último encontro, estava pensando que:
- Design first API apesar de não ser avançado, é para resolver um problema que a galera ainda não tem, não passou
- Estou pensando em ou falar disto no final ou talvez deixar para uma segunda rodada, pode acelerar e ter mais mão na massa e quem sabe até um código mais enxuto
- Assim, acredito que podemos sim colocar o FastAPI no lugar do Flask
- Já brinquei um pouco com FastAPI, mas nunca usei em PROD igual já fiz com Flask, vocês topam usar FastAPI mesmo assim?

Valeu

**Reply by Riverfount** [reply]
Eu topo usar fastAPI, pois poderei ajudar mais. Inclusive minha palestra na Python Brasil será sobre o fastAPI!

**Reply by huogerac** [reply]
Fechou então!
Valeu

**Reply by Riverfount** [reply]
Eu topo usar fastAPI, pois poderei ajudar mais. Inclusive minha palestra na Python Brasil será sobre o fastAPI!

**Reply by huogerac** [reply]
Fechou então!
Valeu

**Reply by huogerac** [reply]
Até tentei River, mas tive que voltar e vamos ficar no Flask mesmo, depois vou dar mais detalhes pq o "ganhar a documentação da API" pode também ser desvantagem.
